### PR TITLE
pfedhpo femnist experiment

### PR DESCRIPTION
**Pull request by TheSunWillRise**

### CODE REVIEW

File: federatedscope/autotune/pfedhpo/fl_server.py [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+        else:
+            for k, v in self._ss.items():
+                if not (hasattr(v, 'lower') and hasattr(v, 'upper')):
+                    if hasattr(v, 'choices'):
+                        self.pbounds[k] = list(v.choices)
+                    else:
+                        raise ValueError("Unsupported hyper type {}".format(type(v)))
+                else:
+                    if v.log:
+                        l, u = np.log10(v.lower), np.log10(v.upper)
```
Comment (joneswong) [inline federatedscope/autotune/pfedhpo/fl_server.py:102]: the lower and upper bounds of `v` are always positive, right?

File: federatedscope/autotune/pfedhpo/fl_server.py [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+                                            ).to(self._cfg.device)
+
+            self.HyperNet.load_state_dict(hyper_enc['hyperNet'])
+            self.HyperNet.eval()
+            if not self.discrete:
+                self.raw_params = self.HyperNet()[0].detach().cpu().numpy()
+            else:
+                self.logits = self.HyperNet()[0]
+
+    def callback_funcs_model_para(self, message: Message):
```
Comment (joneswong) [inline federatedscope/autotune/pfedhpo/fl_server.py:147]: is it necessary to have this branch: https://github.com/alibaba/FederatedScope/blob/b3f823d138e574b070704017b934fb1fdb2750d5/federatedscope/core/workers/server.py#L900

**Review by joneswong — CHANGES_REQUESTED** [review]
There are many hardcoded behaviors. We'd better make this module more general.

File: federatedscope/core/workers/server.py [hunk 3]
```diff
@@ -646,8 +646,11 @@ def broadcast_model_para(self,
                     timestamp=self.cur_timestamp,
                     content=model_para))
         if self._cfg.federate.online_aggr:
-            for idx in range(self.model_num):
-                self.aggregators[idx].reset()
+            try:
+                for idx in range(self.model_num):
+                    self.aggregators[idx].reset()
+            except:
```
Comment (joneswong) [inline federatedscope/core/workers/server.py:652]: we should not ignore exceptions happening here

File: federatedscope/core/optimizers/adadp.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
+import torch
+from torch.optim.optimizer import Optimizer
+import numpy as np
+
+class ADADP(Optimizer):
```
Comment (joneswong) [inline federatedscope/core/optimizers/adadp.py:5]: I didn't see any usage of this optimizer. Which module has instantiated ADADP?

File: federatedscope/core/configs/cfg_fl_setting.py [hunk 5]
```diff
@@ -140,15 +140,15 @@ def assert_fl_setting_cfg(cfg):
...
-            if sample_client_num_valid:
-                logger.warning(
-                    f"Users specify both valid sample_client_rate as"
-                    f" {cfg.federate.sample_client_rate} "
-                    f"and sample_client_num as {old_sample_client_num}.\n"
-                    f"\t\tWe will use the sample_client_rate value to "
-                    f"calculate "
-                    f"the actual number of participated clients as"
-                    f" {cfg.federate.sample_client_num}.")
+            # if sample_client_num_valid:
```
Comment (joneswong) [inline federatedscope/core/configs/cfg_fl_setting.py:143]: any difference?

File: federatedscope/autotune/pfedhpo/fl_server.py (hunk 1 above)
Comment (TheSunWillRise) [reply federatedscope/autotune/pfedhpo/fl_server.py:82]: yes

File: federatedscope/autotune/pfedhpo/fl_server.py (hunk 2 above)
Comment (TheSunWillRise) [reply federatedscope/autotune/pfedhpo/fl_server.py:129]: not necessary

File: federatedscope/core/optimizers/adadp.py (hunk 4 above)
Comment (TheSunWillRise) [reply federatedscope/core/optimizers/adadp.py:5]: This is an implementation of the following paper:

A. Koskela and A. Honkela. "Learning Rate Adaptation for Differentially Private Learning." In: International Conference on Artificial Intelligence and Statistics. PMLR, 2020. p. 2465-2475.
